Add docs for events and actions fetched from network API (#312)
o1-labs/docs2: docs/zkapps/advanced-snarkyjs/events.mdx

- Events are not stored on-chain. Only events from the most recent couple of transactions are retained by consensus nodes. After that, they will be discarded, but are still accessible on archive nodes. In the near future, we plan to add an API to easily fetch events from an archive node.
- You can't refer to previously emitted events in a smart contract, because there is no way of proving that the events you refer to are actually the events emitted by that contract.

:::tip

A zkApp can retrieve events and actions from one or more Mina archive nodes. If your smart contract needs to fetch events and actions from an archive node, see [How to Fetch Events and Actions](fetch-events-and-actions).

:::

> This is all you need to know about events! Think of them as a convenience feature -- a light-weight way of attaching information about your smart contract execution, which would otherwise get lost. Don't treat them as fully-fledged storage that can be safely accessed in smart contracts.

Reviewer: should we use an actual em dash here? `—`

Reviewer: Suggestion. 'Getting lost' seems negative and I wonder if this suggestion gets to the intent better:

`, which would otherwise get lost`   >  `, to expose the event to the outside world such as your UI`

}
```

### Fetching Events from an Archive Node

> Since emitted event data is only stored in transaction data, events will only be persisted by an Archive Node. To fetch events from an Archive Node, you can use the `archive` property on the `Mina.Network` function. For example:

Reviewer: Would suggest rewording the first sentence:

> Events are contained within zkApp transactions and persisted by Mina archive nodes.

Would reword the 2nd sentence similar to my suggestions in the earlier section.

```ts

> let Network = Mina.Network({

Reviewer: const

  mina: '', // Use https://proxy.berkeley.minaexplorer.com/graphql or https://api.minascan.io/node/berkeley/v1/graphql
  archive: '', // Use https://api.minascan.io/archive/berkeley/v1/graphql/ or https://archive.berkeley.minaexplorer.com/
});
Mina.setActiveInstance(Network);
```

> By setting the `archive` property to one of the listed endpoints, SnarkyJS will use one of the selected endpoints to fetch events. For example, let us assume that a zkApp has already emitted events and has broadcasted the transaction. To fetch events for a zkApp, you can use the `fetchEvents` function:

Reviewer: Would rephrase b/c the URL isn't the actor. I.e. it's fetched from an endpiont, the endpoint doesn't do it.

`fetchEvents()`

> By setting the `archive` property to one of the listed endpoints, SnarkyJS will use one of the selected endpoints to fetch events. For example, let us assume that a zkApp has already emitted events and has broadcasted the transaction. To fetch events for a zkApp, you can use the `fetchEvents` function:

Reviewer 2: ```suggestion
By setting the `archive` property to one of the listed endpoints, SnarkyJS uses one of the selected endpoints to fetch events. For example, if a zkApp has already emitted events and has broadcasted the transaction. To fetch events for a zkApp, you can use the `fetchEvents` function:
```

```ts
let zkapp = new SimpleZkapp(address);
// fetches all events from zkapp starting block 0

> let events = await zkapp.fetchEvents(UInt32.from(0));

Reviewer: Could comment out one of the lines and start both with `let`, since it's an alternate example and both lines won't be used simultaneously.

fetches > fetch

// fetches all events starting at block 560 and ending at block 600
events = await zkapp.fetchEvents(UInt32.from(560), UInt32.from(600));
// fetches all events for a given address
const fetchedEvents = await fetchEvents({
  publicKey: 'B62qrfn5xxChtPGJne9HuDJZ4ziWVgWxeL3hntGBqMmf45p4hudo3tw',
});
```

> `getEvents` takes in a starting block and an ending block. If the ending block is not provided, it will fetch all events up to the latest block. If the starting block is not provided, it will fetch all events from the beginning of the zkApp's history.

Reviewer: should be `fetchEvents()`, not `getEvents()` right?

> `getEvents` takes in a starting block and an ending block. If the ending block is not provided, it will fetch all events up to the latest block. If the starting block is not provided, it will fetch all events from the beginning of the zkApp's history.

Reviewer 2: ```suggestion
`getEvents` takes in a starting block and an ending block. 

- If the ending block is not provided, it fetches all events up to the latest block. 
- If the starting block is not provided, it fetches all events from the beginning of the zkApp's history.
```

}
```

### Fetching Events from an Archive Node

Since Mina nodes do not store historical network information and events are not kept on-chain, emitted events are preserved only in the archive node. To see how to work with events within your zkApp, refer to our [Using the Archive Node with SnarkyJS](../../node-operators/archive-node#using-the-archive-node-with-snarkyjs) section.

> If your smart contract needs to fetch events from an archive node, simply provide an `archive` property within the object passed into `Mina.Network({mina: '...', archive: '...'})`. Its value should be the URL for the GraphQL API service that you want to use. If this property does not exist, an error will occur.

Reviewer 2: same feedback as above... we can work together to find the best way forward to omit duplicate architectural content

// Fetch all events starting at block 560 and ending at block 600
const events = await zkapp.fetchEvents(UInt32.from(560), UInt32.from(600));

// Fetch all events for a given address
const fetchedEvents = await fetchEvents({
  publicKey: 'B62qrfn5xxChtPGJne9HuDJZ4ziWVgWxeL3hntGBqMmf45p4hudo3tw',
});
```

By default, `fetchEvents()` retrieves all events from the very first ever emitted for that zkApp account. Additionally, you can provide a starting block height and an ending block height as optional parameters to limit the range of events fetched. If the ending block is not provided, `fetchEvents()` fetches all events up to the latest block, and if the starting block is not provided, it fetches all events from the beginning of the zkApp's history.

To see a end-to-end example of zkApp fetching events from a running network, see this voting app example in the [run_berkeley.ts](https://github.com/o1-labs/snarkyjs/blob/main/src/examples/zkapps/voting/run_berkeley.ts) file in examples provided in the codebase repo.

### Events vs Actions

Events and [Actions](actions-and-reducer) are two distinct mechanisms for logging information alongside a transaction.

> 

Reviewer: These could be added to the glossary. Actions & events are features of Mina Protocol.

}
```

> ### Fetching Events from an Archive Node

Reviewer: My suggestions above can be looked at and adapted down here too, if accepted, given this section mirrors the above.

Reviewer 2: @jasongitmail  see the updated PR description about having a single source of the archive node content... as best practice, we do not want to mirror or duplicate content

> Since Mina nodes do not store historical network information and events are not kept on-chain, emitted events are only preserved in the Archive Node. Consequently, a zkApp must fetch its previously emitted events on the Mina network from an Archive Node.

Reviewer 2: ```suggestion
Since Mina nodes do not store historical network information and events are not kept on-chain, emitted events are preserved only in the Archive Node. Consequently, a zkApp must fetch its previously emitted events on the Mina network from an Archive Node.
```
